File: 114-flatten-binary-tree-to-linked-list/flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        if (root == null) return;
        flattenUtil(root);
    }
    public TreeNode flattenUtil(TreeNode root) {
        if (root == null) return null;
        if (root.left == null && root.right == null) return root;
        TreeNode rootLeft = root.left;
        TreeNode rootRight = root.right;
        TreeNode lastLeftNode = flattenUtil(rootLeft);
        TreeNode lastRightNode = flattenUtil(rootRight);
        if (lastLeftNode == null) {
            // don't need to do anything
            return lastRightNode;
        } else if (lastRightNode == null) {
            root.right = rootLeft;
            root.left = null;
            return lastLeftNode;
        } else {
            lastLeftNode.right = rootRight;
            root.left = null;
            root.right = rootLeft;
            return lastRightNode;
        }
    }
}

File: 114-flatten-binary-tree-to-linked-list/flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if root == None:
            return
        dummy_root = TreeNode(-1)
        head = dummy_root
        def traverse(curr_node):
            nonlocal dummy_root
            if curr_node == None:
                return None
            left_node = curr_node.left
            right_node = curr_node.right
            dummy_root.right = curr_node
            curr_node.left = curr_node.right = None
            dummy_root = dummy_root.right
            # go on the left subtree
            traverse(left_node)
            traverse(right_node)
        traverse(root)
        return head.right